### PR TITLE
Add support for Alpine guest OS

### PR DESCRIPTION
**Pull request by jandubois**
Requires Alpine image with cloud-init, but configures lima-guestagent service with openrc and pretends /bin/ash can stand in for /bin/bash.

Does not work with containerd; not sure where to put the OpenRC scripts for it (maybe nerdctl?).

###### *Persistent disk when booting from ISO image*

When booting from a LiveCD image there is no point in generating a `diffDisk` because the OS is running from RAM, not from the disk.

In that case just add `diffDisk` as an independent data volume and move directories that should be persisted on there. Currently: `/home`, `/usr/local`, `/etc/containerd`, and `/var/lib/containerd`.

This `data-volume` logic is currently only enabled for Alpine because it hasn't been tested anywhere else yet.

Additional, if the configured disk size is 0, no `diffDisk` will be allocated and changes will be written back to the `baseDisk` (or discarded, if the `baseDisk` is an ISO image).

### This is still WIP

Things seem to be working fine, except `limactl stop` so far doesn't manage to shut down Alpine gracefully.

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Thanks! But why not use openSUSE? 😛

File: pkg/iso9660util/iso9660util.go [hunk 1]
```diff
@@ -60,3 +60,18 @@ func WriteFile(fs filesystem.FileSystem, path string, r io.Reader) (int64, error
 	defer f.Close()
 	return io.Copy(f, r)
 }
+
+func IsISO9660(imagePath string) (bool, error) {
```
Comment (AkihiroSuda) [inline pkg/iso9660util/iso9660util.go:64]: Fast path: `if strings.HasSuffix(strings.ToLower(imagePath), ".iso") { return true, nil }`

File: pkg/qemu/qemu.go [hunk 2]
```diff
@@ -61,14 +62,19 @@ func EnsureDisk(cfg Config) error {
 		}
 	}
 	diskSize, err := units.RAMInBytes(cfg.LimaYAML.Disk)
+	if err != nil || diskSize == 0 {
+		return err
```
Comment (AkihiroSuda) [inline pkg/qemu/qemu.go:66]: return a new error when `diskSize == 0 && err == nil`(does it happen? it should be already validated)

File: pkg/qemu/qemu.go [hunk 3]
```diff
@@ -118,11 +124,28 @@ func Cmdline(cfg Config) (string, []string, error) {
...
 		logrus.Warnf("field `firmware.legacyBIOS` is not supported for architecture %q, ignoring", y.Arch)
 	}
-	args = append(args, "-boot", "order=c,splash-time=0,menu=on")
-
-	// Root disk
-	args = append(args, "-drive", fmt.Sprintf("file=%s,if=virtio", filepath.Join(cfg.InstanceDir, filenames.DiffDisk)))
 
+	baseDisk := filepath.Join(cfg.InstanceDir, filenames.BaseDisk)
+	diffDisk := filepath.Join(cfg.InstanceDir, filenames.DiffDisk)
+	isCDROM, err := iso9660util.IsISO9660(baseDisk)
```
Comment (AkihiroSuda) [inline pkg/qemu/qemu.go:130]: `isCDROM` -> `IsBaseDiskCDROM` for readability

File: pkg/cidata/user-data.TEMPLATE [hunk 4]
```diff
@@ -91,6 +219,15 @@ write_files:
         {{- if .Containerd.User}}
         dnf install -y shadow-utils
         {{- end}}
+      elif command -v apk 2>&1 >/dev/null; then
+        : {{/* make sure the "elif" block is never empty */}}
+        {{- if .Mounts}}
+        if ! command -v sshfs 2>&1 >/dev/null; then
+          apk update
+          apk add sshfs
```
Comment (AkihiroSuda) [inline pkg/cidata/user-data.TEMPLATE:227]: apk update is probably not needed

File: pkg/cidata/user-data.TEMPLATE [hunk 5]
```diff
@@ -18,6 +18,129 @@ users:
     {{- end}}
 
 write_files:
```
Comment (AkihiroSuda) [inline pkg/cidata/user-data.TEMPLATE:20]: The `user-data.TEMPLATE` is becoming too long.
Eventually we should just place the scripts as plain files in the `cidata` drive.
(And we can execute scripts from the `cidata` drive directly, without copying them into `/var/lib/lima-guestagent`.)

Can be another PR, though.

Comment (AkihiroSuda) [reply pkg/cidata/user-data.TEMPLATE:20]: And in future we should also have shellcheck in CI

File: pkg/cidata/user-data.TEMPLATE [hunk 6]
```diff
@@ -63,11 +186,16 @@ write_files:
       umount /mnt/lima-cidata
 
       # Launch the guestagent service
-      until [ -e "/run/user/{{.UID}}/systemd/private" ]; do sleep 3; done
-      sudo -iu "{{.User}}" "XDG_RUNTIME_DIR=/run/user/{{.UID}}" lima-guestagent install-systemd
+      if [ -f /etc/alpine-release ]; then
+        rc-update add lima-guestagent default
+        rc-service lima-guestagent start
```
Comment (AkihiroSuda) [inline pkg/cidata/user-data.TEMPLATE:191]: On systemd host, we have `lima-guestagent install-systemd`, so maybe we should have `lima-guestagent install-openrc`?

File: pkg/cidata/user-data.TEMPLATE [hunk 7]
```diff
@@ -99,7 +236,7 @@ write_files:
       fi
       {{- end}}
    owner: root:root
-   path: /var/lib/cloud/scripts/per-boot/10-install-packages.boot.sh
+   path: /var/lib/cloud/scripts/per-boot/30-install-packages.boot.sh
    permissions: '0755'
 {{- end}}
 {{- if or .Containerd.System .Containerd.User}}
```
Comment (AkihiroSuda) [inline pkg/cidata/user-data.TEMPLATE:242]: `.Containerd.System` and `.Containerd.User` should be ignored with a warning message on non-systemd hosts, until we can have openrc scripts

**Comment by jandubois** [comment]
> Thanks! But why not use openSUSE? 😛

I'm looking at replacing the macOS VM management in [Rancher Desktop](https://rancherdesktop.io/) with lima (the first alpha used a hacked up minikube build with a k3s bootstrapper, the current version uses docker-machine-driver-hyperkit with a custom control program).

For that I want the smallest possible image that can run k3s, which means little more than a kernel with support for `/dev`, `/proc`, and `/sys`, an init process, and `/bin/sh`. On Windows we rely on WSL, so our distro is just

```console
$ ls -lh distro.tar
-rw-r--r--  1 jan  staff   3.3M 10 Mar 17:29 distro.tar
$ tar tvfz distro.tar | grep -v ^[ld]
-rwxr-xr-x  0 root   root   303088 17 Feb 17:23 bin/find
-rwxr-xr-x  0 root   root   812632 17 Feb 17:23 bin/busybox
-rwxr-xr-x  0 root   root   610884 10 Mar 17:29 bin/screen
-rwxr-xr-x  0 root   root  1376352 17 Feb 17:23 bin/coreutils
-rw-r--r--  0 root   root       30 10 Mar 17:29 etc/passwd
-rw-r--r--  0 root   root   214057 22 Feb 11:51 etc/ssl/certs/ca-certificates.crt
-rw-r--r--  0 root   root     3810 10 Mar 17:29 usr/share/terminfo/x/xterm-256color
-rwxr-xr-x  0 root   root      428 10 Mar 17:29 usr/local/bin/kubeconfig
-rwxr-xr-x  0 root   root     1437 10 Mar 17:29 usr/local/bin/run-k3s
```

From that `bin/find`, `bin/screen`, and the terminfo entry aren't even used anymore.

The reason to slim down the distro is so that we are not affected by CVEs in software that we don't even run. Having a 500MB Linux distro means we have to update it frequently, the updates will be huge, and the users not happy...

Obviously we'll need more than 3MB on macOS, but I think 40-50MB instead of 550MB is a reasonable target. The "virt" image for Alpine is 41MB, and grows to 61MB when you add cloud-init. This obviously is also ridiculous, and should be replaced by a small bash script instead. 😄 

We don't want to bother with having the OS update itself; we just want to replace the OS and keep the kube objects and container images on a separate volume. This means using an ISO image is ideal, as it is also much smaller than the corresponding QCOW2 image.

There is another project/experiment building OS images from containers: [rancher-sandbox/cOS-toolkit](https://github.com/rancher-sandbox/cOS-toolkit). I will challenge them to produce an openSUSE based ISO that is at most 50% larger than my Alpine image. :) https://github.com/rancher-sandbox/cOS-toolkit/issues/26

Anyways, this is probably much more than you wanted to know. 😄

File: pkg/cidata/user-data.TEMPLATE (hunk 5 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:20]: shellcheck becomes much easier once the scripts are individually embedded templates; I just don't know if it will choke on the template expressions, like `{{.User}}` in the scripts.

There is also potentially a size limit on `user-data`, but I think the limit comes from the cloud-provider side, and not from cloud-init, so it probably doesn't apply to lima. But splitting out the files makes sense anyways.

File: pkg/cidata/user-data.TEMPLATE (hunk 6 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:191]: I actually started that way, and may even still have the branch here locally, but then realized that including the file in `user-data` is much easier.

Added complexity is that the OpenRC script must be installed by `root`, and `lima-guestagent` doesn't allow running as `root`, so that logic also became more complicated.

Is there a reason not to generate the `systemd` config via template in `user-data` (or as a separate file) as well?

File: pkg/cidata/user-data.TEMPLATE (hunk 4 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:227]: It is needed, at least in my image, which only includes a static apkovl in the repositories by default.

It is still 🚧, but the script I use to build (various different) ISOs is at https://github.com/rancher-sandbox/alpine-lima

File: pkg/cidata/user-data.TEMPLATE (hunk 7 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:242]: Yes, but the "warning message" will only be in `/var/log/cloud-init-output.log` on the guest. We don't know if the OS is using systemd or openrc at the time the template is rendered, and I intentionally didn't want to add another config setting for it.

I guess I could add a "requirement" for it though, to check once the VM is up and running.

File: pkg/qemu/qemu.go (hunk 2 above)
Comment (jandubois) [reply pkg/qemu/qemu.go:66]: I think `err` will always be `nil` at this point, so the logic really is:

```go
	diskSize, _ := units.RAMInBytes(cfg.LimaYAML.Disk)
	if diskSize == 0 {
		return nil
        }
```

Maybe it is clearer, but then I would always wonder why it is safe to ignore the error... 😄

File: pkg/cidata/user-data.TEMPLATE (hunk 6 above)
Comment (AkihiroSuda) [reply pkg/cidata/user-data.TEMPLATE:191]: I think we can modify the guest agent to allow running as the root.

> Is there a reason not to generate the systemd config via template in user-data (or as a separate file) as well?

No specific reason

File: pkg/iso9660util/iso9660util.go (hunk 1 above)
Comment (jandubois) [reply pkg/iso9660util/iso9660util.go]: Sounds good, I switched to that one. Also cannot fail. 😄

File: pkg/qemu/qemu.go (hunk 2 above)
Comment (jandubois) [reply pkg/qemu/qemu.go]: Confirmed that this call cannot fail, so ignoring error return value

File: pkg/cidata/user-data.TEMPLATE (hunk 5 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:20]: > Can be another PR, though.

Yes, this PR is complex enough. Let me know if you expect me to do it; or if you will do it as part of your regular refactoring...

File: pkg/cidata/user-data.TEMPLATE (hunk 6 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:213]: > > Is there a reason not to generate the systemd config via template in user-data (or as a separate file) as well?
> 
> No specific reason

In that case I would want to simplify the systemd config in a separate commit, once the user-data refactoring is done.

**Comment by jandubois** [comment]
> `limactl stop` so far doesn't manage to shut down Alpine gracefully.

Turns out `acpid` was not running because the image was based on the `virt` flavour.

File: pkg/cidata/user-data.TEMPLATE (hunk 5 above)
Comment (AkihiroSuda) [reply pkg/cidata/user-data.TEMPLATE:20]: I'll do

File: pkg/cidata/user-data.TEMPLATE (hunk 7 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:285]: I've tried to implement this in d981c4f but am not really happy with it:

1. The `fatal` field seems like a kludge, but there is no point in retrying the check for `systemd`; it isn't going to magically appear eventually.
2. The failure output is really ugly/hard to read. I suspect all requirements failures look like this now though, so maybe not something to address in this PR?

```
INFO[0035] [hostagent] Waiting for the optional requirement 1 of 2: "systemd must be available"
INFO[0035] [hostagent] Forwarding "/run/user/501/lima-guestagent.sock" (guest) to "/Users/jan/.lima/lima/ga.sock" (host)
INFO[0035] [hostagent] No further optional requirements will be checked
ERRO[0035] [1 error occurred:
	* failed to satisfy the optional requirement 1 of 2 "systemd must be available": systemd is required to run containerd, but does not seem to be available.
Make sure that you use an image that supports systemd. If you do not want to run
containerd, please make sure that both 'container.system' and 'containerd.user'
are set to 'false' in the config file.
; skipping further checks: stdout="", stderr="+ command -v systemctl\n+ echo 'systemd is not available on this OS'\nsystemd is not available on this OS\n+ exit 1\n": failed to execute script "systemd must be available": stdout="", stderr="+ command -v systemctl\n+ echo 'systemd is not available on this OS'\nsystemd is not available on this OS\n+ exit 1\n": exit status 1

]
WARN[0035] DEGRADED. The VM seems running, but file sharing and port forwarding may not work. (hint: see "/Users/jan/.lima/lima/ha.stderr.log")
FATA[0035] degraded, status={Running:true Degraded:true Exiting:false Errors:[1 error occurred:
	* failed to satisfy the optional requirement 1 of 2 "systemd must be available": systemd is required to run containerd, but does not seem to be available.
Make sure that you use an image that supports systemd. If you do not want to run
containerd, please make sure that both 'container.system' and 'containerd.user'
are set to 'false' in the config file.
; skipping further checks: stdout="", stderr="+ command -v systemctl\n+ echo 'systemd is not available on this OS'\nsystemd is not available on this OS\n+ exit 1\n": failed to execute script "systemd must be available": stdout="", stderr="+ command -v systemctl\n+ echo 'systemd is not available on this OS'\nsystemd is not available on this OS\n+ exit 1\n": exit status 1

] SSHLocalPort:60022}
```

Suggestions welcome!

File: pkg/cidata/user-data.TEMPLATE [hunk 8]
```diff
@@ -18,11 +18,141 @@ users:
...
+      test -f /etc/alpine-release || exit 0
+
+      # Redirect bash to ash (built with CONFIG_ASH_BASH_COMPAT) and hope for the best :)
+      # It does support `set -o pipefail`, but not `[[`.
+      # /bin/bash can't be a symlink because /bin/ash is a symlink to /bin/busybox
+      cat >/bin/bash <<'EOF'
+      #!/bin/sh
+      exec /bin/ash "$@"
+      EOF
+      chmod +x /bin/bash
```
Comment (AkihiroSuda) [inline pkg/cidata/user-data.TEMPLATE:34]: Why not just `if [ ! -e /bin/bash ]; then ln -s ash /bin/bash; fi`

Comment (AkihiroSuda) [reply pkg/cidata/user-data.TEMPLATE:34]: Also curious, do we really need bash?

Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:34]: See the comment in line 29:

```
# /bin/bash can't be a symlink because /bin/ash is a symlink to /bin/busybox
```

busybox does not implement `bash`:

```console
lima-lima:/bin# busybox bash
bash: applet not found
```

So this cannot work:

```console
lima-lima:/bin# rm /bin/bash
lima-lima:/bin# ln -s /bin/ash /bin/bash
lima-lima:/bin# bash -c uname
bash: applet not found
```

Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:34]: > Also curious, do we really need bash?

Normal `/bin/sh` does not support `set -o pipefail`, so we couldn't use that in any of the scripts. `/bin/ash` supports it because it is compiled in bash-compatibility mode.

I added the `/bin/bash` wrapper because I didn't want to templatize the `#!/bin/bash` line in all the scripts. That would also require the user to specify the shell in the limayaml because we don't know if the image has `/bin/bash` until it is running, at which time cloud-config has already been processed.

I generally try to avoid adding new config options if reasonably possible; they will eventually become unwieldy even when you practise restraint. 😄

**Comment by jandubois** [comment]
@AkihiroSuda I've switched the PR to non-draft now because I think it works as-is.

I will need to spend some more effort on the alpine-lima repo to document how to create images compatible with lima etc. Not sure if you want to wait for that, or review/merge now and then update later with e.g. an `example/alpine.yaml`?

**Comment by AkihiroSuda** [comment]
Thanks. The doc can be added later, but could you add the example YAML so that I can test it?

Would be great if we can test it on CI, too. https://github.com/AkihiroSuda/lima/blob/36b2ab60cd8ebf18bd232833c78e3603aae953c4/.github/workflows/test.yml#L55

File: pkg/iso9660util/iso9660util.go (hunk 1 above)
Comment (jandubois) [reply pkg/iso9660util/iso9660util.go:64]: I had to revert that change because I call it on "baseDisk", which has no file extension.

Anyways, even the "proper" implementation is pretty fast, just opening a file and reading a 32KB chunk; this should not be noticeable compare to everything else that is happening.

**Review by AkihiroSuda — APPROVED** [review]
Thanks!

File: pkg/cidata/user-data.TEMPLATE (hunk 4 above)
Comment (jandubois) [reply pkg/cidata/user-data.TEMPLATE:271]: When I tested it again right now, it seemed like `apk update` is indeed not needed. Not sure which image I was testing when I did need it.  I can remove it, if you prefer, or leave it in. It doesn't really hurt, as `apk add sshfs` will still update the repos anyways, so I would prefer to leave it as-is.
